k2/csrc/array_ops.cu

```diff
@@ -147,63 +148,55 @@ Array1<int32_t> SpliceRowSplits(int32_t num_arrays,
 
 bool ValidateRowIds(Array1<int32_t> &row_ids, Array1<int32_t> *temp) {
   ContextPtr ctx = row_ids.Context();
-  int32_t *data = row_ids.Data();
+  const int32_t *data = row_ids.Data();
   int32_t dim = row_ids.Dim();
-  if (dim == 0) return true;  // will treat this as valid.a
-
-  if (ctx->GetDeviceType() == kCpu) {
-    if (data[0] < 0) return false;
-    for (int32_t i = 0; i + 1 < dim; i++)
-      if (data[i] > data[i + 1]) return false;
-  }
+  if (dim == 0) return true;  // will treat this as valid
+  // note `row_ids[0]` may copy memory from device to host
+  if (row_ids[0] < 0) return false;
 
   Array1<int32_t> temp_array;
   if (temp == nullptr || temp->Dim() == 0) {
-    temp_array = Array1<int32_t>(row_ids.Context(), 1);
-    temp = &temp_array;
+    temp_array = Array1<int32_t>(ctx, 1);
+  } else {
+    temp_array = temp->Range(0, 1);
   }
-
+  temp = &temp_array;
   *temp = 0;
+
   int32_t *temp_data = temp->Data();
   auto lambda_check_row_ids = [=] __host__ __device__(int32_t i) -> void {
-    int32_t this_val = data[i], next_val = data[i + 1];
-    if (this_val > next_val || this_val < 0) *temp_data = 1;  // means it's bad.
+    if (data[i] > data[i + 1]) *temp_data = 1;  // means it's bad.
```

Comment:
All row ids should be less than `dim`.

Reply:
No, such row ids would be valid, [0,0,1,8], i.e. there are a lot of empty rows.